docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,151 @@
+
+```
+Query 1--LSIs: {a, b}, reduced_cost: 4
+Query 2--LSIs: {c},    reduced_cost: 4
+Query 3--LSIs: {a, d}, reduced_cost: 3
+```
+- **Step2**: get the score list. Simply add up `LSIs`' `reduced_cost` of each query respectively and then sort the list descending by score. Assume this step outputs:
+
+```
+a: 4+3=7
+b: 4
+c: 4
+d: 3
+```
+
+- **Step3**: evaluate recommend index set. Assume the number of allowed indexes recommended by index advisor is *2* and the swap size is *1* 
+  - The initial `recommend index set` is `{a, b}`, the `remaining index set` is `{c,d}` 
+  - Add recommend index set `{a, b}` to the schema and at the same time mute existed index `c`
+  - Compile all queries in the workload and add up each query's cost. Assume the sum is *20* and then assign it variable `min_cost`
+  - Randomly exchange *1* element between the recommend set and remaining set. Assume *a* and *c* is chosen respectively, now the recommend set is `{c, b}` and the remaining set is `{a, d}`. 
+  - Apply recommend index set `{c, b}` to the workload and get a sum cost, say *18*. 
+  - As the cost decreased from *20* to *18*, update the recommend set from `{a, b}` to `{c, b}` and update `min_cost` to *18*. Then try swapping 1 element between `{c, b}` and `{a, d}` and execute the re-evaluate process. 
+
+Note that if the sum cost with `{c, b}` is *22* ( greater than 20 ), the next *swap* process should happen between `{a, b}` and `{c, d}` instead of between `{c, b}` and `{a, d}`. Loop the *Swap and Re-evaluate* process until timeout.
+
+## Rational 
+
+In this design, *index advisor* is implemented in a user friendly way, by adding a new SQL syntax `index advise...` instead of an independent third-party tool. However, for simplicity, syntax like `add virtual index...` which serves as building blocks of *index advisor* is not exposed. Therefore, if DBAs want some functionality like `add virtual index` or `alter invisible index`, they cannot benefit directly from current *index advisor* implementation. 
+
+## Compatibility
+
+No compatibility issues.
+
+## Implementation
+
+### <span id="compatibility">**Support from TiDB SQL optimizer** </span>
+
+1. **add hypothetic indexes to schema** 
+   - **effect**: query should be compiled with *schema* that contains input hypothetic indexes 
+   - **constraints**: these hypothetic indexes should take effects only within current index advisor session  
+   - **implementation**: get a copy of *session schema* in memory. Then add hypothetic indexes to schema copy and compile the query with this modified schema copy
+
+2. **alter invisible indexes**
+   - **effect**: input existed indexes on table should be invisible to optimizer 
+   - **constraints**: only within current index advisor session, input existed indexes are invisible to optimizer 
```

Comment:
```suggestion
   - **constraints**: only within current index advisor session, inputting existing indexes are invisible to optimizer 
```

Reply:
Or do you mean `existing input indexes`?

Reply:
I want to express two staffs:
1. Users are asked to type in some `indexes`. These indexes are the `input` of index advisor.
2. And these indexes must already be `existing` on tables.